include/class_loader/class_loader.hpp

```diff
@@ -122,12 +122,21 @@ class ClassLoader
    * @return A std::shared_ptr<Base> to newly created plugin object
    */
   template<class Base>
-  std::shared_ptr<Base> createInstance(const std::string & derived_class_name)
+  std::shared_ptr<Base> createInstance(
+    const std::string & derived_class_name,
+    bool is_shared_ptr = false)
```

Comment:
@ahcorde meta: I understand we do this for API compatibility, but all downstream code that invokes this directly is likely  leaking (unless it manages to establish ownership on its own). Not sure what to do though, this is an old library with lots of users. CC'ing @clalancette @wjwwood @mjcarroll for input.

Reply:
So I'll say that I'm not really familiar with the class_loader API at the moment.

But isn't this an internal implementation detail?  That is, we are constructing a `std::shared_ptr` with a custom deleter (`onPluginDeletion`), and whether we pass it a shared_ptr or a regular pointer is a detail.  From that perspective, it seems like we should always use the `shared_ptr` version so we properly clean up later on.

I guess you could argue that some downstream consumers may be expecting that we don't properly clean up, but I'd argue that that is a bug in the downstream consumers.

If we *do* decide to get rid of the non-shared_ptr version here, I think it would do us well to do an audit of the downstream consumers to see what they expect here.  Outside of the tests, there are only a handful of uses in the ROS 2 core.  It might also be interesting to look at the wider group of packages that are released into Foxy and Galactic, just to see what the usage patterns look like there.

Reply:
what's about this other idea https://github.com/ros/class_loader/pull/186/commits/6a800a44a6613fa8c4de88d7efbb07363006f52c ?

Using a `try-catch` block ?

Reply:
With the `try/catch` approach all downstream packages will benefit from these changes without making any modification in the code.

Reply:
Try-catch seems like a good idea. We can just document that if you do not create it as a shared pointer then you have to ensure that it outlives the objects created from it.